### ISSUE TITLE
Loss-based unsupervised skill recovery

### ISSUE THREAD

**Issue by G-JWLee** [HazyResearch/skill-it#1]
Hello, thank you for your great work!

The recovery of skills using the loss trajectory is really interesting and intuitive.
I want to do the research based on your findings.

I'm curious about the precise steps of the skill recovery procedure based on loss observation.
Here is what I understood:

(1) Train the model and measure the evaluation loss with the evaluation dataset for 120 checkpoints.
(2) Repeat this for 10 runs, which results is Nx1200 loss matrix where N is the number of evaluation samples.
(3) Use the Nx1200 observation for k-means clustering, which ends in k clusters, which is the identified skill set.

However, after this process, we don't know in which cluster the training data should be assigned, hence cannot apply the skillit method based on the identified skill during training.

Could you help me understand the process?
Thank you in advance!

**Comment by mayeechen** [comment]
Hi, thanks for checking out our work! 

The skill recovery algorithm mentioned in the paper was not used end-to-end in training and serves more as a proof-of-concept of a method whose recovered clusters on validation data line up more with the synthetic skills than e.g., embedding clustering. 

To directly extend this to training data, one possibility is to just take the training dataset (call it $D_{train}$) and split it into the training dataset ($D_{train}^1$) for skill recovery, and the held out training dataset ($D_{train}^2$). Then, run the skill recovery approach by training on $D_{train}^1$ and evaluating on $D_{train}^2$, such that your matrix input to k-means is of size $|D_{train}^2| \times 1200$. Unfortunately this method results in only being able to use the clustered $D_{train}^2$ moving forward. 

Another idea is to simply train and evaluate on all of $D_{train}$ for skill recovery, although I'm not sure how well that works.

A final idea that I have been exploring is using gradient clustering rather than this approach. That is; take your base model and get the gradient of each sample in the training dataset. Then normalize and perform spectral clustering to get the recovered skills. This might work better and is hopefully simpler! I took inspiration from https://github.com/ejmichaud/quanta-discovery/tree/dev (paper: https://arxiv.org/abs/2303.13506) for this.

Let me know if you have any more questions!

**Comment by G-JWLee** [comment]
Thank you for your kind reply! The first method can only cluster the $D_{train}^2$, which is the subset of the full train dataset $D_{train}$. Then, did you split the full dataset into N splits, and repeat the process over N times to fill out the $|D_{train}| \times 1200$ loss trajectory matrix? For example, you can split the full dataset into 4 splits, use one of the split subsets for evaluation and use the other three subsets for training. After measuring the $|D_{train}|/4 \times 1200$ from the previous process, change the evaluation subset into another subset and again train the model with the other three subsets for training, and repeat this to fill out the $|D_{train}| \times 1200$ matrix. Is this the method that you used in your paper?

**Comment by mayeechen** [comment]
That is a good idea and I think it makes sense to try out! The method we used in our paper was purely on validation data, e.g. we had a $D_{train}$ and $D_{val}$ and just used the clustering algorithm to look at the recovered skills on $D_{val}$. The purpose of our experiment was to measure how aligned these clusters were with the true synthetic skills, and we did not use these clusters as input to the actual skill-it method. I hope this helps!
